### ISSUE TITLE
[AOTAutograd] cuDNN BatchNorm decomposition to native BatchNorm changes training numerics

### ISSUE THREAD

**Issue by whitephilomel** [pytorch/pytorch#187444]
### 🐛 Describe the bug

I see a reproducible numerical divergence between eager DDP and `torch.compile` DDP when running the same model, optimizer, seed, synthetic input tensors, and two-node NCCL setup. Both arms complete 20 optimizer steps without NaN/Inf, but the final logits fingerprint and parameter sum differ substantially.

This is not caused by dataset ordering: the reproducer uses fixed-seed synthetic tensors. AMP is disabled, TF32 is disabled, cuDNN benchmark is disabled, deterministic mode is enabled, and `CUBLAS_WORKSPACE_CONFIG=:4096:8` is used. The config contains `"precision": "bf16"`, but effective compute is fp32 because `amp_enabled=false`.

Minimal reproducer and configuration:

https://gist.github.com/whitephilomel/13d0584f7396760793b11dbbe7491dfd

## Current behavior

With PyTorch 2.9.1+cu128, two machines, one V100 GPU per machine, `world_size=2`, backend `nccl`, model `resnet50`, optimizer `adamw`, lr `1.0`, and 20 steps:

| rank | eager loss | compile loss | loss abs diff | eager grad norm | compile grad norm | grad abs diff | divergent checks |
|---|---:|---:|---:|---:|---:|---:|---|
| rank 0 | 2.3456020355 | 3.6093058586 | 1.2637038231 | 0.8535786867 | 1.0606021881 | 0.2070235014 | logits norm, logits sum, parameter sum |
| rank 1 | 2.7591028214 | 2.5354700089 | 0.2236328125 | 0.8535786867 | 1.0606021881 | 0.2070235014 | logits norm, logits sum, parameter sum |

The largest differences are in the final logits and parameter fingerprint:

| metric | eager | compile | abs diff |
|---|---:|---:|---:|
| rank0 logits norm | 17.5366973877 | 34.8463401794 | 17.3096427917 |
| rank0 logits sum | 149.2867736816 | 300.2042541504 | 150.9174804688 |
| rank0 parameter sum | 379786.78125 | -98813.453125 | 478600.234375 |
| rank1 logits norm | 17.5704326630 | 34.7212409973 | 17.1508083344 |
| rank1 logits sum | 149.4028778076 | 299.1468811035 | 149.7440032959 |
| rank1 parameter sum | 379786.78125 | -98813.453125 | 478600.234375 |

## Expected behavior

For the same DDP world size, model, optimizer, initialization seed, and deterministic synthetic inputs, eager and `torch.compile` should either produce numerically close results within a reasonable tolerance or fail consistently. A large compile-only change in final logits and parameter fingerprints after successful training steps is unexpected.

### Versions

PyTorch: 2.9.1+cu128
Python: 3.11.15
CUDA reported by PyTorch: 12.8
NCCL observed in confirmed runs: 2.27.5+cuda12.9
OS/platform: Linux 5.15.0-113-generic x86_64, glibc 2.35
GPU: one Tesla V100S-PCIE-32GB per machine
Distributed setup: 2 machines, world_size=2, one process per machine

cc @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste @ezyang @gchanan @kadeng @msaroufim @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @aditvenk @weifengpy @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @azahed98

**Comment by xtLeo2023** [comment]
I reproduced this issue with the provided repro script. Have you been looking further into the root cause? Maybe you can run it on a single node to check whether the problem persists.

**Comment by shino16** [comment]
Inductor uses different algorithm/reduction-order choices for convolutions and batch normalization vs eager, which produces ~1e-2 max logit difference even on a single ResNet50 forward pass. With `lr=1.0` the training is numerically unstable, and I think it's not really a bug.

**Comment by krastogi-in** [comment]
I agree with @shino16 and I see all Five GitHub issues ([#187444](https://github.com/pytorch/pytorch/issues/187444), [#187445](https://github.com/pytorch/pytorch/issues/187445), [#187446](https://github.com/pytorch/pytorch/issues/187446), [#187447](https://github.com/pytorch/pytorch/issues/187447), [#187448](https://github.com/pytorch/pytorch/issues/187448)) reported that ResNet models trained with torch.compile + DDP produce silently different loss, gradients, and logits compared to eager mode are similar . You can even get the any one of them without ddp and backend to be inductor.

**Reason**

Inductor decomposes native_batch_norm_backward into Triton kernels that sum thousand of float32 values in a different tree order than ATen's C++ kernel. Both are mathematically correct, but float32 addition is non-associative: (a+b)+c != a+(b+c). With multiple additions, the accumulated rounding difference is ~1e-4 per element. This compounds through 20+ BN layers per backward pass, through skip connections, and across training steps — especially with Adam.

**Fix Options**

1. Accept the divergence — this is inherent to compiling reductions into GPU kernels with different accumulation order. Eager and compiled training should reach similar final accuracy, just via different trajectories.
2. Keep native_batch_norm_backward as an extern kernel in Inductor — similar to how convolution_backward is already handled. Trades fusion opportunities for bit-exact parity.
3. Use SGD or lower Adam's learning rate to reduce amplification. (Workaround.)

I think priority of the issue can be lowered as this is expected behavior. Or go with 2nd option.

**Comment by eellison** [comment]
I investigated this together with #187445–#187449.

The compiler bisector localizes #187444, #187445, #187447, and #187449 to AOTAutograd:

```text
BisectionResult(backend='aot_eager', subsystem=None)
```

This identifies the AOTAutograd stage, not a particular decomposition. Profiling and graph inspection show the relevant path change:

```text
aten.cudnn_batch_norm          -> aten.native_batch_norm
aten.cudnn_batch_norm_backward -> aten.native_batch_norm_backward
```

In direct deterministic tests, each implementation was bitwise repeatable, while cuDNN versus native differed by approximately `5e-8`–`1.4e-7` relative per operation. Native-BatchNorm eager then matched ordinary `aot_eager` bitwise, including outputs and parameter gradients, for all four configurations. This confirms that the initial difference comes from the BatchNorm implementation change, rather than Inductor, DDP, NCCL, or the optimizer.

The larger endpoint differences arise as those small numerical differences are amplified through the model and training steps. The same underlying behavior was previously discussed in #93787.

I am keeping this issue as the canonical AOTAutograd/cuDNN-versus-native BatchNorm numerical-parity tracker. I am closing #187445, #187447, and #187449 as duplicates. #187446 and #187448 are being closed separately because their endpoint predicates also fail plain eager-versus-eager controls.
